### PR TITLE
Integration gateway profiler

### PR DESCRIPTION
**Pull request by petergrlica**
**What the feature does**

Adds a service and apigw subsystem that carries info on:
 - missed hits on registered routes
 - missed hits on non-existing routes
 - aggregations + hit details per-route, per-hit

**Please check**
Is the service and aggregation on the right layer (service)?
Can sorting be optimized?

Note, there is no hit purging yet, so the performance implications on that area are known.

### CODE REVIEW

File: system/types/app_settings.go [hunk 1]
```diff
@@ -207,6 +207,12 @@ type (
 			Enabled bool `kv:"-" json:"enabled"`
 		} `kv:"federation" json:"federation"`
 
+		// Federation settings
```
Comment (darh) [inline system/types/app_settings.go]: leftover.

File: pkg/apigw/registry/registry.go [hunk 2]
```diff
@@ -8,19 +8,22 @@ import (
...
-		h map[string]types.Handler
+		opts *options.ApigwOpt
+		h    map[string]types.Handler
 	}
 
 	secureStorageTodo struct{}
 )
 
-func NewRegistry() *Registry {
+func NewRegistry(opts *options.ApigwOpt) *Registry {
```
Comment (darh) [inline pkg/apigw/registry/registry.go]: Options should b (unwritten Corteza rule) passed to the functions as values (not refs).

File: pkg/apigw/helpers.go [hunk 3]
```diff
@@ -21,3 +25,35 @@ func helperDefaultResponse(opt *options.ApigwOpt) http.HandlerFunc {
...
 	}
 }
+
+func helperMethodNotAllowed(opt *options.ApigwOpt, pr *profiler.Profiler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		addToProfiler(opt, pr, r, http.StatusMethodNotAllowed)
+
+		if opt.LogEnabled {
+			// Say something friendly when logging is enabled
+			http.Error(w, devHelperResponseBody, http.StatusTeapot)
```
Comment (darh) [inline pkg/apigw/helpers.go:37]: This is nice :)
But why not properly returning 405 even when logging? The fact that logging is enabled does not affect (other) behaviour of the handler and its responses , correct?

File: pkg/apigw/helpers.go [hunk 4]
```diff
@@ -21,3 +25,35 @@ func helperDefaultResponse(opt *options.ApigwOpt) http.HandlerFunc {
...
+func addToProfiler(opt *options.ApigwOpt, pr *profiler.Profiler, r *http.Request, status int) {
+	if !(opt.ProfilerEnabled && opt.ProfilerGlobal) {
+		return
+	}
+
+	// add to profiler
+	ar, err := h.NewRequest(r)
+
+	if err != nil {
+		panic(err)
```
Comment (darh) [inline pkg/apigw/helpers.go]: Can this situation be caused by a non-developer (so anyone w/o access to server source)?
If yes, then this should be converted to a log (zap).

File: pkg/apigw/profiler/hit.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	h "github.com/cortezaproject/corteza-server/pkg/http"
+)
+
+type (
+	Hit struct {
```
Comment (darh) [inline pkg/apigw/profiler/hit.go:12]: what was the rationale for abbreviating fields in this struct?

File: app/options/apigw.cue [hunk 6]
```diff
@@ -25,6 +25,14 @@ apigw: schema.#optionsGroup & {
 			type:        "bool"
 			description: "Enable extra logging"
 		}
+		profiler_enabled: {
+			type:        "bool"
+			description: "Enable profiler"
+		}
+		profiler_global: {
+			type:        "bool"
+			description: "Profiler enabled for all routes"
```
Comment (darh) [inline app/options/apigw.cue]: do we need to explain in more detail gow enabling this affects profilers on routes?

File: app/options/apigw.cue [hunk 7]
```diff
@@ -25,6 +25,14 @@ apigw: schema.#optionsGroup & {
 			type:        "bool"
 			description: "Enable extra logging"
 		}
+		profiler_enabled: {
```
Comment (darh) [inline app/options/apigw.cue:28]: should this be enabled by default?

File: pkg/apigw/ctx/ctx.go [hunk 8]
```diff
@@ -3,12 +3,14 @@ package ctx
...
 	"context"
 
+	"github.com/cortezaproject/corteza-server/pkg/apigw/profiler"
 	"github.com/cortezaproject/corteza-server/pkg/apigw/types"
 )
 
 type ContextKey string
 
 const ContextKeyScope ContextKey = "scope"
+const ContextKeyProfiler ContextKey = "profiler"
```
Comment (darh) [inline pkg/apigw/ctx/ctx.go]: can you change this (both?) to follow "empty-struct" approach (see `identityCtxKey`) we use for context key?

File: app/options/apigw.cue (hunk 6 above)
Comment (darh) [reply app/options/apigw.cue]: and, should we log this into stdout (warn/info/debug) that global profiler is enabled?

**Review by darh — CHANGES_REQUESTED** [review]

File: pkg/apigw/ctx/ctx.go (hunk 8 above)
Comment (petergrlica) [reply pkg/apigw/ctx/ctx.go]: :+1: agree

File: app/options/apigw.cue (hunk 7 above)
Comment (petergrlica) [reply app/options/apigw.cue:28]: > should this be enabled by default?

It should be on by default and global off, will act accordingly

File: app/options/apigw.cue (hunk 6 above)
Comment (petergrlica) [reply app/options/apigw.cue]: > do we need to explain in more detail gow enabling this affects profilers on routes?

I'll add a little more info on server logs, most will be in docs

> and, should we log this into stdout (warn/info/debug) that global profiler is enabled?

yes

File: pkg/apigw/profiler/hit.go (hunk 5 above)
Comment (petergrlica) [reply pkg/apigw/profiler/hit.go:12]: None other than self-explanatory

File: pkg/apigw/helpers.go (hunk 4 above)
Comment (petergrlica) [reply pkg/apigw/helpers.go]: Did not want to wrap with logger and the rationale was that something really wrong should go when creating a wrapper request from `http.Request`, but there could be some issues with ie headers or body, so I'll revert that panic, :+1:

File: pkg/apigw/helpers.go (hunk 3 above)
Comment (petergrlica) [reply pkg/apigw/helpers.go:37]: I just wanted a teapot, I agree with you though

File: pkg/apigw/registry/registry.go (hunk 2 above)
Comment (petergrlica) [reply pkg/apigw/registry/registry.go]: :+1:

**Pull request merged**